File: tools/lib/merkle.mjs
/**
 * merkle.mjs — a domain-separated binary Merkle tree over brick seals.
 *
 * One Merkle ROOT commits to every brick's provenance seal at once, so a single
 * external anchor (Sigstore/Rekor, OpenTimestamps, Solana, …) timestamps the
 * whole ledger. Any individual brick is then proven against that root with a
 * small inclusion proof — no need to anchor 3,148 things, and no need to reveal
 * any other brick (privacy: proofs disclose sibling hashes, never content).
 *
 * Domain separation: leaves are prefixed 'leaf\0', internal nodes 'node\0', so
 * a node hash can never be reinterpreted as a leaf (second-preimage defense).
 * Odd layers duplicate the last node (Bitcoin-style). Pure + deterministic.
 */

import { createHash } from 'node:crypto';

const sha256 = (s) => createHash('sha256').update(s).digest('hex');

/** Leaf hash binds a brick id to its seal head. */
export function leafHash(brickId, sealHead) {
  return sha256(`leaf\0${brickId || ''}\0${sealHead || ''}`);
}

function nodeHash(a, b) {
  return sha256(`node\0${a}\0${b}`);
}

/**
 * Build a Merkle tree from ordered leaf hashes.
 * Returns { root, layers } where layers[0] === leaves and the last layer is [root].
 */
export function buildMerkle(leaves) {
  if (!leaves || !leaves.length) return { root: sha256('empty\0'), layers: [[]] };
  const layers = [leaves.slice()];
  while (layers[layers.length - 1].length > 1) {
    const prev = layers[layers.length - 1];
    const next = [];
    for (let i = 0; i < prev.length; i += 2) {
      const a = prev[i];
      const b = i + 1 < prev.length ? prev[i + 1] : prev[i]; // duplicate last if odd
      next.push(nodeHash(a, b));
    }
    layers.push(next);
  }
  return { root: layers[layers.length - 1][0], layers };
}

/**
 * Inclusion proof for the leaf at `index`: an ordered list of
 * { hash, side } siblings from leaf up to (but excluding) the root.
 */
export function inclusionProof(layers, index) {
  const proof = [];
  let idx = index;
  for (let l = 0; l < layers.length - 1; l += 1) {
    const layer = layers[l];
    const isRight = idx % 2 === 1;
    const sibIdx = isRight ? idx - 1 : (idx + 1 < layer.length ? idx + 1 : idx);
    proof.push({ hash: layer[sibIdx], side: isRight ? 'left' : 'right' });
    idx = Math.floor(idx / 2);
  }
  return proof;
}

/** Low-level: verify a leaf + proof reproduce the root. Callers MUST have
 *  derived `leaf` themselves via leafHash — passing an untrusted leaf lets a
 *  caller "prove" an internal node. Prefer verifyBrickInclusion below. */
export function verifyProof(leaf, proof, root) {
  let h = leaf;
  for (const step of proof) {
    h = step.side === 'left' ? nodeHash(step.hash, h) : nodeHash(h, step.hash);
  }
  return h === root;
}

/** Safe inclusion check: re-derives the leaf from (brickId, sealHead) so an
 *  attacker cannot substitute an internal-node hash as the "leaf". This is the
 *  entry point external verifiers should use. */
export function verifyBrickInclusion(brickId, sealHead, proof, root) {
  return verifyProof(leafHash(brickId, sealHead), proof, root);
}

export { sha256 as _sha256 };

File: rust-core/src/merkle.rs
use serde::Serialize;
use sha2::{Digest, Sha256};

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct ProofStep {
    pub hash: String,
    pub side: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MerkleTree {
    pub root: String,
    pub layers: Vec<Vec<String>>,
}

fn sha256(text: &str) -> String {
    format!("{:x}", Sha256::digest(text.as_bytes()))
}

pub fn leaf_hash(brick_id: &str, seal_head: &str) -> String {
    sha256(&format!("leaf\0{brick_id}\0{seal_head}"))
}

fn node_hash(left: &str, right: &str) -> String {
    sha256(&format!("node\0{left}\0{right}"))
}

pub fn build_merkle(leaves: &[String]) -> MerkleTree {
    if leaves.is_empty() {
        return MerkleTree {
            root: sha256("empty\0"),
            layers: vec![Vec::new()],
        };
    }
    let mut layers = vec![leaves.to_vec()];
    while layers.last().is_some_and(|layer| layer.len() > 1) {
        let previous = layers.last().expect("non-empty layer");
        let next = previous
            .chunks(2)
            .map(|pair| node_hash(&pair[0], pair.get(1).unwrap_or(&pair[0])))
            .collect();
        layers.push(next);
    }
    MerkleTree {
        root: layers.last().unwrap()[0].clone(),
        layers,
    }
}

pub fn inclusion_proof(layers: &[Vec<String>], index: usize) -> Vec<ProofStep> {
    let mut proof = Vec::new();
    let mut index = index;
    for layer in layers.iter().take(layers.len().saturating_sub(1)) {
        let is_right = index % 2 == 1;
        let sibling = if is_right {
            index - 1
        } else if index + 1 < layer.len() {
            index + 1
        } else {
            index
        };
        proof.push(ProofStep {
            hash: layer[sibling].clone(),
            side: if is_right { "left" } else { "right" }.into(),
        });
        index /= 2;
    }
    proof
}

pub fn verify_proof(leaf: &str, proof: &[ProofStep], root: &str) -> bool {
    proof.iter().fold(leaf.to_owned(), |hash, step| {
        if step.side == "left" {
            node_hash(&step.hash, &hash)
        } else {
            node_hash(&hash, &step.hash)
        }
    }) == root
}

pub fn verify_brick_inclusion(
    brick_id: &str,
    seal_head: &str,
    proof: &[ProofStep],
    root: &str,
) -> bool {
    verify_proof(&leaf_hash(brick_id, seal_head), proof, root)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn duplicates_odd_leaf_and_verifies_proofs() {
        let leaves: Vec<String> = (0..5)
            .map(|index| leaf_hash(&format!("brick-{index}"), &format!("head-{index}")))
            .collect();
        let tree = build_merkle(&leaves);
        for (index, leaf) in leaves.iter().enumerate() {
            assert!(verify_proof(
                leaf,
                &inclusion_proof(&tree.layers, index),
                &tree.root
            ));
        }
        assert!(!verify_brick_inclusion(
            "brick-4",
            "wrong",
            &inclusion_proof(&tree.layers, 4),
            &tree.root
        ));
    }
}
